src/Sparrow/Json/ArenaMemoryAllocator.cs

```diff
@@ -0,0 +1,124 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using NLog;
+
+namespace Sparrow.Json
+{
+    public unsafe class ArenaMemoryAllocator : IDisposable
+    {
+        private byte* _ptrStart;
+        private byte* _ptrCurrent;
+
+        private int _allocated;
+        private int _used;
+
+        private List<IntPtr> _unusedBuffers;
+
+        private bool _isDisposed;
+        private static readonly Logger _log = LogManager.GetLogger(nameof(ArenaMemoryAllocator));
+        
+        public ArenaMemoryAllocator(int initialSize = 1024 * 1024)
+        {
+            _ptrStart = _ptrCurrent = (byte*)Marshal.AllocHGlobal(initialSize).ToPointer();
+            _allocated = initialSize;
+            _used = 0;
+
+            _log.Debug($"ArenaMemoryAllocator was created with initial capacity of {initialSize:#,#;;0} bytes");
+        }
+
+        public AllocatedMemoryData Allocate(int size)
+        {
+            if (_isDisposed)
+                ThrowAlreadyDisposedException();
+
+            if (_used + size > _allocated)
+                GrowArena(size);
+
+            var allocation = new AllocatedMemoryData()
+            {
+                SizeInBytes = size,
+                Address = new IntPtr(_ptrCurrent)
+            };
+
+            _ptrCurrent += size;
+            _used += size;
+
+            _log.Debug($"ArenaMemoryAllocator allocated {size:#,#;;0} bytes");
+
+            return allocation;
+        }
+
+        private void ThrowAlreadyDisposedException()
+        {
+            throw new ObjectDisposedException($"This ArenaMemoryAllocator is already disposed");
+        }
+
+        public void GrowArena(int requestedSize)
+        {
+            while (_used + requestedSize > _allocated)
+            {
+                var newSize = _allocated * 2;
+                if(newSize < _allocated)
+                    throw new OverflowException("Arena size overflowed");
+
+                _log.Debug($"ArenaMemoryAllocator doubled size of buffer from {_allocated:#,#;0} to {newSize:#,#;0}");
+                _allocated = newSize;
+            }
+            
+            var newBuffer = (byte*) Marshal.AllocHGlobal(_allocated).ToPointer();
+            
```

Comment:
You can use Marshal.ReallocHGlobal() to grow an arena. It could fail if there is no contiguous memory available but that will help to keep fragmentation even more in check.